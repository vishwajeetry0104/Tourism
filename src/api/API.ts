import axios, {AxiosInstance, AxiosRequestConfig} from 'axios';

class API {
  API: AxiosInstance | undefined;
  constructor() {
    this.init();
  }

  async init() {
    const config = {
      ajaxSettings: {
        headers: {
          'Cache-Control': 'no-cache',
          'Content-Type': 'application/json',
          Accept: 'application/json',
        },
        timeout: 45000,
      },
      env: {
        apiUrl: 'https://web-dev.dev.kimo.ai',
      },
    };

    const axiosOptions = {
      baseURL: config.env.apiUrl,
      timeout: config.ajaxSettings.timeout,
      headers: config.ajaxSettings.headers,
    };
    this.API = axios.create(axiosOptions);
  }

  async get(url: string, headerData?: AxiosRequestConfig<{}>): Promise<any> {
    if (this.API) {
      return this.API.get(`${url}`, headerData);
    }
  }

  async post(url: string, data: {}) {
    return this.API?.post(`${url}`, {
      ...data,
    });
  }
}

const api = new API();
export default api;
