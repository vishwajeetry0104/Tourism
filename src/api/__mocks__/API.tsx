import API from '../API';

const mockedAPI = jest.fn().mockImplementation(() =>
  Promise.resolve({
    data: {},
  }),
);

API.get = mockedAPI;
API.post = mockedAPI;

module.exports = {
  __esModule: true,
  default: API,
  get: mockedAPI,
  post: mockedAPI,
};
