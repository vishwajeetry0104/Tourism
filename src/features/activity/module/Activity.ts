import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
import type {PayloadAction} from '@reduxjs/toolkit';
import API from '../../../api/API';

type Activity = {
  name: string
}

type ActivityDetail = {
  name: string,
  description: string,
  image: string,
  activities: Array<Activity>
}

interface InitialStateType {
  loading: boolean;
  activityTypes: {
    'Surfing': ActivityDetail,
    'Traditional Festivals': ActivityDetail,
    'Volcanoes': ActivityDetail
  };
  hasFailed: boolean;
}

const defaultActivityDetail = {
  name: '',
  description: '',
  image: '',
  activities: []
};

export const initialState: InitialStateType = {
  loading: false,
  activityTypes: {
    'Surfing': defaultActivityDetail,
    'Traditional Festivals': defaultActivityDetail,
    'Volcanoes': defaultActivityDetail
  },
  hasFailed: false
};

export const fetchActivityDetail = createAsyncThunk('activity/details', async ({activityType}: {activityType: string}) => {
  const response = await API.get(`/v1/activities/${activityType}`);
  return response?.data;
});

export const activitySlice = createSlice({
  name: 'activity',
  initialState,
  reducers: {
    resetActivity: (state: InitialStateType) => {
      state.activityTypes = {
        'Surfing': defaultActivityDetail,
        'Traditional Festivals': defaultActivityDetail,
        'Volcanoes': defaultActivityDetail
      };
      state.loading = false;
      state.hasFailed = false;
    },
  },
  extraReducers: builder => {
    builder.addCase(
      fetchActivityDetail.fulfilled,
      (state: InitialStateType, action: PayloadAction<ActivityDetail, any, {arg: {activityType: string}}>) => {
        const {activityType} = action?.meta?.arg ?? {activityType: ''};
        state.activityTypes[activityType] = action.payload;
        state.loading = false;
        state.hasFailed = false;
      },
    );
    builder.addCase(fetchActivityDetail.pending, (state: InitialStateType) => {
      state.loading = true;
      state.hasFailed = false;
    });
    builder.addCase(fetchActivityDetail.rejected, (state: InitialStateType) => {
      state.loading = false;
      state.hasFailed = true;
    });
  },
});

export const {resetActivity} = activitySlice.actions;

export default activitySlice.reducer;
