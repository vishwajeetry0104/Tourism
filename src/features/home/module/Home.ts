import {createSlice, createAsyncThunk} from '@reduxjs/toolkit';
import type {PayloadAction} from '@reduxjs/toolkit';
import API from '../../../api/API';

interface InitialStateType {
  loading: boolean;
  data: [];
  hasFailed: boolean;
}

export const initialState: InitialStateType = {
  loading: false,
  data: [],
  hasFailed: false,
};

export const fetchHighlights = createAsyncThunk('home/highlights', async () => {
  const response = await API.get('/highlights');
  return response?.data;
});

export const homeSlice = createSlice({
  name: 'home',
  initialState,
  reducers: {
    resetHome: (state: InitialStateType) => {
      state.data = [];
      state.loading = false;
      state.hasFailed = false;
    },
  },
  extraReducers: builder => {
    builder.addCase(
      fetchHighlights.fulfilled,
      (state: InitialStateType, action: PayloadAction<any>) => {
        state.loading = false;
        state.hasFailed = false;
      },
    );
    builder.addCase(fetchHighlights.pending, (state: InitialStateType) => {
      state.loading = true;
      state.hasFailed = false;
    });
    builder.addCase(fetchHighlights.rejected, (state: InitialStateType) => {
      state.loading = false;
      state.hasFailed = true;
    });
  },
});

export const {resetHome} = homeSlice.actions;

export default homeSlice.reducer;
