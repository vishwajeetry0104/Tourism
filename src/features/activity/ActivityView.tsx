import React, {useEffect} from 'react';
import {useAppDispatch, useAppSelector} from "../../store/Hooks";
import {RootState} from "../../store/Store";
import { ImageBackground, useWindowDimensions, Text, View, ScrollView, Alert } from "react-native";
import CategoryItem from "../../components/categoryItem/CategoryItem";
import TappableButton from "../../components/button/TappableButton";
import TravelGuideCard from "../../components/travelGuide/TravelGuideCard";
import {fetchActivityDetail, resetActivity, ActivityType} from "./module/Activity";
import Spinner from "../../components/spinner/Spinner";
import InfoCard from "../../components/infoCard/InfoCard";
import Styles from './ActivityViewStyles';
import GradientText from "../../components/gradientText/GradientText";
import PlaceHolderImage from '../../assets/images/placeholder.png';

type Props = {
  activityType: ActivityType
}

const ActivityView = ({activityType}: Props) => {
  const {width} = useWindowDimensions();
  const dispatch = useAppDispatch();

  const {
    activityTypes, loading, hasFailed
  } = useAppSelector((state: RootState) => state.activity);

  useEffect(() => {
    dispatch(fetchActivityDetail({activityType}));
    return () => {
      dispatch(resetActivity());
    }
  }, []);

  const onPressCategory = (name) => {
    Alert.alert(`${name} pressed`);
  }

  return (
    <>
      <ScrollView style={Styles.scrollViewContainer} contentContainerStyle={Styles.contentContainerStyle}>
        {hasFailed ? (
          <View style={[Styles.flatListEmptyComponentContainer, {backgroundColor: 'transparent'}]}>
            <InfoCard
              testID="infoCard_activity"
              headingTitle='No results'
              message={hasFailed ? 'Something went wrong': 'Nothing to display'}
            />
          </View>
        ) : (
          <>
            <ImageBackground source={activityTypes[activityType]?.image ? {uri: activityTypes[activityType]?.image} : PlaceHolderImage} style={{height: (width / 3) * 2, width: width}} resizeMode="cover">
              <GradientText colors={['rgba(255, 255, 255, 0.20)', 'rgba(255, 255, 255, 0.90)']} style={Styles.imageBackgroundText}>
                {activityTypes[activityType]?.name}
              </GradientText>
            </ImageBackground>
            <Text style={Styles.descriptionText}>{activityTypes[activityType]?.description}</Text>
            <Text style={[Styles.sectionLabelText, {backgroundColor: 'transparent'}]}>Top spots</Text>
            {activityTypes[activityType]?.activities.map((item) => {
              return (
                <CategoryItem
                  testID={item.name}
                  key={item.name}
                  name={item.name}
                  onPressCategory={() => onPressCategory(item.name)}
                />
              )
            })}
          </>
        )}
        <TravelGuideCard name="Hadwin Malone" description="Guide since 2012" onContactPress={() => {}} />
      </ScrollView>
      <Spinner testID="loader" animating={loading} subTitle={activityType} />
      <View style={Styles.floatingButtonContainer}>
        <TappableButton title="Book a trip" onButtonPress={() => {}} />
      </View>
    </>
  );
};

export default ActivityView;
