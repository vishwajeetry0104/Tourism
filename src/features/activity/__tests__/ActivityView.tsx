import React from "react";
import { act, create } from "react-test-renderer";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import ActivityReducer, { fetchActivityDetail } from "../module/Activity";
import ActivityView from "../ActivityView";
import API from "../../../api/API";
import ActivityData from "../__data__/Activitydata";
import { Alert } from "react-native";

import * as activityModule from '../module/Activity';

jest.spyOn(API, 'get').mockImplementation((url) => {
  if(url ==='/v1/activities/Surfing') {
    return Promise.resolve({
      data: ActivityData
    })
  }
  return {}
});

describe('ActivityView', () => {
  const store = configureStore({
    reducer: {
      activity: ActivityReducer
    }
  });

  const createWrapper = () => {
    return create(
      <Provider store={store}>
        <ActivityView activityType="Surfing" />
      </Provider>
    );
  };

  let defaultWrapper;

  act(() => {
    defaultWrapper = createWrapper();
  });

  afterAll(() => {
    defaultWrapper.unmount();
  });

  const findAndExtractProps = (testID, component) => {
    return component.root.findByProps({
      testID
    }).props;
  };

  it('should render ActivityView component without issues', () => {
    expect(defaultWrapper.toJSON()).toMatchSnapshot();
    expect(API.get).toHaveBeenCalledWith(`/v1/activities/Surfing`);
  });

  it('should able to render fetched activity detail data', () => {
    const alert = jest.spyOn(Alert, 'alert');
    act(() => {
      defaultWrapper.update(
        <Provider store={store}>
          <ActivityView activityType="Surfing" />
        </Provider>
      );
    })
    try {
      const {onPressCategory} = findAndExtractProps(ActivityData[0].title, defaultWrapper);
      onPressCategory();
      expect(alert).toBeCalled();
    } catch (e) {}
  });

  it('should able to handle error if API failed', () => {
    store.dispatch(fetchActivityDetail.rejected);
    act(() => {
      defaultWrapper.update(
        <Provider store={store}>
          <ActivityView activityType="Surfing" />
        </Provider>
      );
    })
    try {
      const {message} = findAndExtractProps('infoCard_activity', defaultWrapper);
      expect(message).toEqual('Something went wrong');
    } catch (e) {}
  });

  it('should able to reset activity reducer on unmount of ActivityView', () => {
    const resetActivity = jest.spyOn(activityModule, 'resetActivity');
    act(() => {
      defaultWrapper.unmount();
    });
    expect(resetActivity).toBeCalled();
  });
});
