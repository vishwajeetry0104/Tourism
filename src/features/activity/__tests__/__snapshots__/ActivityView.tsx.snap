// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityView should render ActivityView component without issues 1`] = `
[
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 100,
      }
    }
    style={
      {
        "backgroundColor": "#E6F2F2",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          {
            "height": 500,
            "width": 750,
          }
        }
      >
        <Image
          resizeMode="cover"
          source={
            {
              "uri": "https://storage.googleapis.com/topics-images/web-dev-images/surfing.jpg",
            }
          }
          style={
            [
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "height": 500,
                "width": 750,
              },
              undefined,
            ]
          }
        />
        <Text
          style={
            [
              {
                "color": "#FFF",
                "fontSize": 40,
                "fontWeight": "700",
                "lineHeight": 56,
                "marginTop": "25%",
                "textAlign": "center",
              },
              {
                "opacity": 0,
              },
            ]
          }
        >
          Surfing
        </Text>
      </View>
      <Text
        style={
          {
            "backgroundColor": "#FFF",
            "color": "#001A1A",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
            "paddingHorizontal": 16,
            "paddingVertical": 40,
          }
        }
      >
        Hawaii is known for its world-famous surf spots and waves that attract surfers from all over the globe. Here are some of the best islands for surfing in Hawaii:
      </Text>
      <Text
        style={
          [
            {
              "backgroundColor": "#FFF",
              "color": "#001A1A",
              "fontSize": 16,
              "fontWeight": "700",
              "lineHeight": 20,
              "padding": 16,
            },
            {
              "backgroundColor": "transparent",
            },
          ]
        }
      >
        Top spots
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 8,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 16,
            "marginVertical": 4,
            "padding": 24,
          }
        }
      >
        <Text
          style={
            {
              "color": "#001A1A",
              "flexWrap": "wrap",
              "fontSize": 16,
              "fontWeight": "400",
              "paddingRight": 8,
              "width": "80%",
            }
          }
        >
          North Shore, Oahu
        </Text>
        <View
          style={{}}
        >
          <SvgMock
            height={16}
            width={16}
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 8,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 16,
            "marginVertical": 4,
            "padding": 24,
          }
        }
      >
        <Text
          style={
            {
              "color": "#001A1A",
              "flexWrap": "wrap",
              "fontSize": 16,
              "fontWeight": "400",
              "paddingRight": 8,
              "width": "80%",
            }
          }
        >
          Waimea Bay, Oahu
        </Text>
        <View
          style={{}}
        >
          <SvgMock
            height={16}
            width={16}
          />
        </View>
      </View>
      <View
        style={
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 8,
            "flex": 1,
            "justifyContent": "center",
            "margin": 16,
            "padding": 24,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View>
            <Text
              style={
                {
                  "color": "#001A1A",
                  "fontSize": 24,
                  "fontWeight": "700",
                  "paddingBottom": 16,
                }
              }
            >
              Hadwin Malone
            </Text>
            <Text
              style={
                {
                  "color": "#001A1A",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 20,
                  "paddingBottom": 8,
                }
              }
            >
              Guide since 2012
            </Text>
          </View>
          <SvgMock
            height={74}
            width={74}
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "flex-start",
              "backgroundColor": "#FFF",
              "borderColor": "#008080",
              "borderRadius": 8,
              "borderWidth": 1,
              "marginTop": 16,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "#008080",
                "fontSize": 16,
                "fontWeight": "700",
                "paddingHorizontal": 24,
                "paddingVertical": 8,
              }
            }
          >
            Contact
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    transparent={true}
    visible={false}
  />,
  <View
    style={
      {
        "bottom": 0,
        "left": 0,
        "paddingBottom": 16,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#008080",
          "borderRadius": 8,
          "elevation": 7,
          "marginHorizontal": 16,
          "paddingBottom": 11,
          "paddingHorizontal": 24,
          "paddingTop": 9,
          "shadowColor": "rgba(7, 56, 56, 0.80)",
          "shadowOffset": {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.29,
          "shadowRadius": 4.65,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "700",
            "lineHeight": 20,
          }
        }
      >
        Book a trip
      </Text>
    </View>
  </View>,
]
`;
